e2e_tests/tests/experiment/test_custom_searcher.py

```diff
@@ -0,0 +1,580 @@
+            f"custom-searcher-asha-test-fail2-{TIMESTAMP}",
+            [
+                "on_validation_completed",
+                "after_save",
+                "after_save",
+                "after_save",
+                "shutdown",
+            ],
+        ),
+        ("noop.yaml", f"custom-searcher-asha-noop-test-{TIMESTAMP}", []),
+        (
+            "noop.yaml",
+            f"custom-searcher-asha-test-noop-fail1-{TIMESTAMP}",
+            [
+                "initial_operations_start",
+                "after_save",
+                "on_validation_completed",
+            ],
+        ),
+    ],
+)
+def test_run_asha_searcher_exp_core_api(
+    config_name: str, exp_name: str, exception_points: List[str]
+) -> None:
+    config = conf.load_config(conf.fixtures_path("custom_searcher/core_api_searcher_asha.yaml"))
+    config["entrypoint"] += " --exp-name " + exp_name
+    config["entrypoint"] += " --config-name " + config_name
+    if len(exception_points) > 0:
+        config["entrypoint"] += " --exception-points " + " ".join(exception_points)
+    config["max_restarts"] = len(exception_points)
+
+    experiment_id = exp.run_basic_test_with_temp_config(
+        config, conf.fixtures_path("custom_searcher"), 1
+    )
+    session = exp.determined_test_session()
+
+    # searcher experiment
+    searcher_exp = bindings.get_GetExperiment(session, experimentId=experiment_id).experiment
+    assert searcher_exp.state == bindings.determinedexperimentv1State.STATE_COMPLETED
+
+    # actual experiment
+    response = bindings.get_GetExperiments(session, name=exp_name)
+    experiments = response.experiments
+    assert len(experiments) == 1
+
+    experiment = experiments[0]
+    assert experiment.numTrials == 16
+
+    response_trials = bindings.get_GetExperimentTrials(session, experimentId=experiment.id).trials
+
+    # 16 trials in rung 1 (#batches = 187)
+    assert sum(t.totalBatchesProcessed >= 187 for t in response_trials) == 16
+    # at least 4 trials in rung 2 (#batches = 750)
+    assert sum(t.totalBatchesProcessed >= 750 for t in response_trials) >= 4
+    # at least 1 trial in rung 3 (#batches = 3000)
+    assert sum(t.totalBatchesProcessed == 3000 for t in response_trials) >= 1
+
+    for trial in response_trials:
+        assert trial.state == bindings.determinedexperimentv1State.STATE_COMPLETED
+
+    # check logs to ensure failures actually happened
+    logs = str(
+        subprocess.check_output(
+            ["det", "-m", conf.make_master_url(), "experiment", "logs", str(experiment_id)]
+        )
+    )
+    failures = logs.count("Max retries exceeded with url: http://dummyurl (Caused by None)")
+    assert failures == len(exception_points)
+
+    # check for resubmitting operations
+    resubmissions = logs.count("determined.searcher: Resubmitting operations for event.id=")
+    assert resubmissions == sum([x == "after_save" for x in exception_points])
+
+
+@pytest.mark.e2e_cpu
+@pytest.mark.parametrize(
```

Comment:
you guys went nuts on these tests :tada: 

I like that a lot, especially for a feature with complex distributed state behaviors.